kbmd/kbmd.c

```diff
@@ -0,0 +1,470 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2019, Joyent, Inc.
+ */
+
+#include <sys/corectl.h>
+#include <sys/list.h>
+#include <sys/resource.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/uuid.h>
+#include <sys/wait.h>
+#include <bunyan.h>
+#include <door.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <libnvpair.h>
+#include <libscf.h>
+#include <libzfs.h>
+#include <paths.h>
+#include <port.h>
+#include <priv.h>
+#include <signal.h>
+#include <smbios.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <synch.h>
+#include <umem.h>
+#include <unistd.h>
+#include "kbmd.h"
+
+#define	KBMD_DEFAULT_PLUGIN_DIR		"/usr/lib/kbmd/plugins"
+#define	DEFAULT_SYSPOOL	"zones"
+
+bunyan_logger_t *blog;
+
+char *sys_pool;
+
+mutex_t g_zfs_lock = ERRORCHECKMUTEX;
+libzfs_handle_t *g_zfs;
+
+mutex_t piv_lock = ERRORCHECKMUTEX;
+SCARDCONTEXT piv_ctx;
+
+uuid_t sys_uuid;
+volatile boolean_t kbmd_quit = B_FALSE;
+
+static int kbmd_daemonize(int);
+static void kbmd_fd_setup(void);
+static int kbmd_dir_setup(void);
+static void kbmd_log_setup(int, bunyan_level_t);
+static void kbmd_cleanup(void);
+static void kbmd_load_smf(int);
+static int kbmd_sys_uuid(uuid_t);
+
+int
+main(int argc, char *argv[])
+{
+	const char *doorpath = KBMD_DOOR_PATH;
+	int dirfd, dfd, errval, c;
+	sigset_t set;
+	struct sigaction act;
+	boolean_t opt_d = B_FALSE;
+
+	while ((c = getopt(argc, argv, "d")) != -1) {
+		switch (c) {
+		case 'd':
+			opt_d = B_TRUE;
+			break;
+		}
+	}
+
+	alloc_init();
+	kbmd_fd_setup();
+	dirfd = kbmd_dir_setup();
+	if (opt_d) {
+		dfd = open("/dev/null", O_RDONLY);
+		if (dfd == -1)
+			err(EXIT_FAILURE, "/dev/null");
+	} else {
+		dfd = kbmd_daemonize(dirfd);
+	}
+
+	/*
+	 * Now in the child (non -d)
+	 */
+
+	kbmd_load_smf(dfd);
+	if (sigfillset(&set) != 0)
+		kbmd_dfatal(dfd, "failed to fill a signal set...");
+
+	if (sigdelset(&set, SIGABRT) != 0)
+		kbmd_dfatal(dfd, "failed to unmask SIGABRT");
+
+	if (sigprocmask(SIG_BLOCK, &set, NULL) != 0)
+		kbmd_dfatal(dfd, "failed to set our door signal mask");
+
```

Comment:
They are undone by line 272 right before it returns -- I believe the intention is to bracket those statements to prevent signals from messing things up during a critical section.  The code is almost verbatim from varpd (which I was told was the 'right' way to do this). There is a bit more that happens in varpd in-between the demonize function returning and re-blocking signals that isn't present here -- just to avoid possible future bugs, I'm thinking maybe add a comment above these statements what sort of potential future things should happen before signals are re-blocked.

Reply:
Actually looking closer.. I'm going to rearrange things here a bit -- I think the key thing is that we want the door server threads that get created to block everything but SIGABRT -- the simplest way is to let them inherit from the thread that calls door_create().. In varpd there were some other threads that also get created where the same setup was wanted, but not (currently) the case here...